File: Top_Interview_Questions/Medium/454. 4Sum II/solution4.py
from typing import List


class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        table = {}
        counter = 0
        for num3 in nums3:
            for num4 in nums4:
                sum = num3 + num4
                if not sum in table:
                    table[sum] = 0
                table[sum] += 1
        for num1 in nums1:
            for num2 in nums2:
                target = 0 - num1 - num2
                if target in table:
                    counter += table[target]
        return counter

File: Top_Interview_Questions/Medium/454. 4Sum II/solution4.cpp
#include <unordered_map>
#include <vector>
using namespace std;

/**
 * 參考解答: 為了要從 solution3 再拔掉一層迴圈，我們將 nums3 也存到 hash map 內
 * 時間: O(n^2), 空間: O(n)
 */

class Solution
{
  public:
    int fourSumCount(vector<int> &nums1, vector<int> &nums2, vector<int> &nums3, vector<int> &nums4)
    {
        unordered_map<int, int> table; // sum of num3 and num4 and their frequency
        int counter = 0;
        for (int i = 0; i < nums3.size(); i++)
        {
            for (int j = 0; j < nums4.size(); j++)
            {
                table[nums3[i] + nums4[j]]++;
            }
        }
        for (int i = 0; i < nums1.size(); i++)
        {
            for (int j = 0; j < nums2.size(); j++)
            {
                counter += table[0 - nums1[i] - nums2[j]];
            }
        }
        return counter;
    }
};